bam_markdup.c

```diff
@@ -467,6 +637,238 @@ static int add_duplicate(khash_t(duplicates) *d_hash, bam1_t *dupe) {
+                }
+
+                return ret;
+            }
+
+            dy = strtol(duplicate + dypos, &end, 10);
+
+            if ((duplicate + dypos) == end) {
+                (*warnings)++;
+
+                if (*warnings <= BMD_WARNING_MAX) {
+                    fprintf(stderr, "[markdup] warning: can not decipher Y coordinate in %s.\n", duplicate);
+                }
+
+                return ret;
+            }
+
+            if (oy > dy) {
+                ydiff = oy - dy;
+            } else {
+                ydiff = dy - oy;
+            }
+
+            if (ydiff <= max_dist) ret = 1;
+        }
+    }
+
+    return ret;
+}
+
+
+static int mark_duplicates(md_param_t *param, khash_t(duplicates) *dup_hash, bam1_t *ori, bam1_t *dup,
+                           long *optical, long *warn) {
+    char dup_type = 0;
+    long incoming_warnings = *warn;
+
+    dup->core.flag |= BAM_FDUP;
+
+    if (param->tag) {
+        if (bam_aux_append(dup, "do", 'Z', strlen(bam_get_qname(ori)) + 1, (uint8_t*)bam_get_qname(ori))) {
+            fprintf(stderr, "[markdup] error: unable to append 'do' tag.\n");
+            return -1;
+        }
+    }
+
+    if (param->opt_dist) { // mark optical duplicates
+        if (optical_duplicate(ori, dup, param->opt_dist, warn)) {
+            bam_aux_append(dup, "dt", 'Z', 3, (const uint8_t *)"SQ");
+            dup_type = 'O';
+            (*optical)++;
+        } else {
+            // not an optical duplicate
+            bam_aux_append(dup, "dt", 'Z', 3, (const uint8_t *)"LB");
+        }
+    }
+
+    if ((*warn == BMD_WARNING_MAX) && (incoming_warnings != *warn)) {
+        fprintf(stderr, "[markdup] warning: %ld decipher read name warnings.  New warnings will not be reported.\n",
+                        *warn);
+    }
+
+    if (param->supp) {
+        if (bam_aux_get(dup, "SA") || (dup->core.flag & BAM_FMUNMAP) || bam_aux_get(dup, "XA")) {
+            char *original = NULL;
+
+            if (param->tag) {
+                original = bam_get_qname(ori);
+            }
+
+            if (add_duplicate(dup_hash, dup, original, dup_type))
+                return -1;
+        }
+    }
+
+    return 0;
+}
+
+
+/* Modified Lander-Waterman equation */
+static inline double lw_equation(double x, double c, double n) {
+    return c / x - 1 + exp(-n / x);
+}
+
+
+/* estimate the library size, based on the Picard code in DuplicationMetrics.java*/
+static unsigned long estimate_library_size(unsigned long read_pairs, unsigned long duplicate_pairs) {
+    unsigned long estimated_size = 0;
+
+    read_pairs /= 2;
+    duplicate_pairs /= 2;
+
+    if (read_pairs && duplicate_pairs) {
+        unsigned long unique_pairs = read_pairs - duplicate_pairs;
+        double m = 1;
+        double M = 100;
+        int i;
+
+        if (lw_equation(m * (double)unique_pairs, (double)unique_pairs, (double)read_pairs) < 0) {
```

Comment:
If unique_pairs is 0, this will result in a division by 0.

Reply:
Well unique_pairs can only be zero if all the pairs are duplicates.  I'm not sure that can ever happen, even if markdup is fed bad data.  Still, I'll check for it just in case.